graylog2-server/src/main/java/org/graylog2/shared/buffers/processors/MessageULIDGenerator.java

```diff
@@ -0,0 +1,83 @@
+/**
+ * This file is part of Graylog.
+ *
+ * Graylog is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Graylog is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Graylog.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.graylog2.shared.buffers.processors;
+
+import com.github.benmanes.caffeine.cache.Cache;
+import com.github.benmanes.caffeine.cache.Caffeine;
+import de.huxhorn.sulky.ulid.ULID;
+import org.graylog2.plugin.Message;
+import org.graylog2.shared.journal.Journal;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.inject.Inject;
+import javax.inject.Singleton;
+
+@Singleton
+public class MessageULIDGenerator {
+    private static final Logger LOG = LoggerFactory.getLogger(MessageULIDGenerator.class);
+
+    private final Cache<String, Integer> sequenceNrCache;
+    private final ULID ulid;
+    private static final long RANDOM_MSB_MASK = 0xFFFFL;
+    private static final int OFFSET_GAP = 500;
+
+    @Inject
+    public MessageULIDGenerator(ULID ulid, Journal journal) {
+        this.ulid = ulid;
+
+        sequenceNrCache = Caffeine.newBuilder()
+                .maximumSize(100)
+                .initialCapacity(100)
+                .build();
+    }
+
+    public Message addULID(Message message) {
```

Comment:
What do you think about making the API of this class a bit more testable? Instead of modifying the Message object, I think it's nicer to just return the generated ID, too.

```diff
diff --git graylog2-server/src/main/java/org/graylog2/shared/buffers/processors/MessageULIDGenerator.java graylog2-server/src/main/java/org/graylog2/shared/buffers/processors/MessageULIDGenerator.java
index 9abc3898e..0fade5802 100644
--- graylog2-server/src/main/java/org/graylog2/shared/buffers/processors/MessageULIDGenerator.java
+++ graylog2-server/src/main/java/org/graylog2/shared/buffers/processors/MessageULIDGenerator.java
@@ -18,9 +18,9 @@ package org.graylog2.shared.buffers.processors;
 
 import com.github.benmanes.caffeine.cache.Cache;
 import com.github.benmanes.caffeine.cache.Caffeine;
+import com.google.common.annotations.VisibleForTesting;
 import de.huxhorn.sulky.ulid.ULID;
 import org.graylog2.plugin.Message;
-import org.graylog2.shared.journal.Journal;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
@@ -37,7 +37,7 @@ public class MessageULIDGenerator {
     private static final int OFFSET_GAP = 500;
 
     @Inject
-    public MessageULIDGenerator(ULID ulid, Journal journal) {
+    public MessageULIDGenerator(ULID ulid) {
         this.ulid = ulid;
 
         sequenceNrCache = Caffeine.newBuilder()
@@ -46,7 +46,12 @@ public class MessageULIDGenerator {
                 .build();
     }
 
-    public Message addULID(Message message) {
+    public String createULID(Message message) {
+        return createULID(message.getSourceInputId(), message.getTimestamp().getMillis(), message.getSequenceNr());
+    }
+
+    @VisibleForTesting
+    String createULID(String inputId, long timestamp, int sequenceNr) {
         // Fill the first 16 bits of the ULIDs random section with a sequence number that reflects the order
         // in which messages were received by an input.
         //
@@ -60,13 +65,10 @@ public class MessageULIDGenerator {
         // This means that the first recorded sequence number can be higher than the one of later messages.
         // To account for this, we simply add a constant (OFFSET_GAP) to prevent negative messageSequenceNrs.
 
-        final long messageTimestamp = message.getTimestamp().getMillis();
-        final int sequenceNr = message.getSequenceNr();
-
-        final String key = message.getSourceInputId() + messageTimestamp;
+        final String key = inputId + timestamp;
         final Integer subtrahend = sequenceNrCache.get(key, k -> sequenceNr);
 
-        final ULID.Value nextUlid = ulid.nextValue(messageTimestamp);
+        final ULID.Value nextUlid = ulid.nextValue(timestamp);
         final long mostSignificantBits = nextUlid.getMostSignificantBits();
         final long leastSignificantBits = nextUlid.getLeastSignificantBits();
 
@@ -77,7 +79,6 @@ public class MessageULIDGenerator {
             messageSequenceNr %= RANDOM_MSB_MASK;
         }
         final ULID.Value sequencedUlid = new ULID.Value(msbWithZeroedRandom + messageSequenceNr, leastSignificantBits);
-        message.addField(Message.FIELD_GL2_MESSAGE_ID, sequencedUlid.toString());
-        return message;
+        return sequencedUlid.toString();
     }
 }
diff --git graylog2-server/src/main/java/org/graylog2/shared/buffers/processors/ProcessBufferProcessor.java graylog2-server/src/main/java/org/graylog2/shared/buffers/processors/ProcessBufferProcessor.java
index 9190d0074..27b589612 100644
--- graylog2-server/src/main/java/org/graylog2/shared/buffers/processors/ProcessBufferProcessor.java
+++ graylog2-server/src/main/java/org/graylog2/shared/buffers/processors/ProcessBufferProcessor.java
@@ -135,7 +135,7 @@ public class ProcessBufferProcessor implements WorkHandler<MessageEvent> {
             // Don't overwrite already existing message IDs.
             // They might already be set if we are receiving messages from a Graylog forwarder
             if (message.getField(Message.FIELD_GL2_MESSAGE_ID) == null) {
-                messageULIDGenerator.addULID(message);
+                message.addField(Message.FIELD_GL2_MESSAGE_ID, messageULIDGenerator.createULID(message));
             }
 
             // The processing time should only be set once all message processors have finished
diff --git graylog2-server/src/test/java/org/graylog2/shared/buffers/processors/MessageULIDGeneratorTest.java graylog2-server/src/test/java/org/graylog2/shared/buffers/processors/MessageULIDGeneratorTest.java
new file mode 100644
index 000000000..eaa8efc77
--- /dev/null
+++ graylog2-server/src/test/java/org/graylog2/shared/buffers/processors/MessageULIDGeneratorTest.java
@@ -0,0 +1,18 @@
+package org.graylog2.shared.buffers.processors;
+
+import de.huxhorn.sulky.ulid.ULID;
+import org.junit.Test;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class MessageULIDGeneratorTest {
+    @Test
+    public void generate() {
+        final MessageULIDGenerator generator = new MessageULIDGenerator(new ULID());
+
+        assertThat(generator.createULID("input1", 0, 0)).startsWith("00000000000001");
+        assertThat(generator.createULID("input1", 0, 65536)).startsWith("00000000000001");
+
+        // TODO: Implement more tests :)
+    }
+}

```